src/duct.py

```diff
@@ -164,62 +166,165 @@ def create_and_parse_args():
-            stdout=subprocess.PIPE,
-            stderr=subprocess.PIPE,
-            preexec_fn=os.setsid,
+    stdout, stderr = prepare_outputs(
+        args.capture_outputs, args.outputs, args.output_prefix
+    )
+    process = subprocess.Popen(
+        [str(args.command)] + args.arguments.copy(),
+        stdout=stdout,
+        stderr=stderr,
+        preexec_fn=os.setsid,
+    )
+    session_id = os.getsid(process.pid)  # Get session ID of the new process
+    report = Report(args.command, session_id)
+    report.collect_environment()
+    report.get_system_info()
+
+    if args.record_types in ["all", "processes-samples"]:
+        monitoring_args = [
+            stdout,
+            stderr,
+            report,
+            process,
+            args.report_interval,
+            args.sample_interval,
+            args.output_prefix,
+        ]
+        monitoring_thread = threading.Thread(
+            target=monitor_process, args=monitoring_args
         )
-        session_id = os.getsid(process.pid)  # Get session ID of the new process
-        report = Report(args.command, session_id)
-        report.collect_environment()
-        report.get_system_info()
-
-        while True:
-            elapsed_time = time.time() - report.start_time
-            report.collect_sample()
-            if elapsed_time >= (report.number + 1) * args.report_interval:
-                aggregated = report.aggregate_samples()
-                for pid, pinfo in aggregated.items():
-                    with open(
-                        f"{args.output_prefix}/{pid}_resource_usage.json", "a"
-                    ) as resource_statistics_log:
-                        pinfo["elapsed_time"] = elapsed_time
-                        resource_statistics_log.write(json.dumps(aggregated))
-                report.number += 1
-
-            if process.poll() is not None:  # the passthrough command has finished
-                break
-            time.sleep(args.sample_interval)
+        monitoring_thread.start()
+        monitoring_thread.join()
 
+    if args.record_types in ["all", "system-summary"]:
         with open(
-            f"{args.output_prefix}/system-report.session-{report.session_id}.json", "a"
+            f"{args.output_prefix}/system-report.session-{report.session_id}.json",
+            "a",
         ) as system_logs:
             report.end_time = time.time()
             report.run_time_seconds = f"{report.end_time - report.start_time}"
             report.get_system_info()
             system_logs.write(str(report))
-        stdout, stderr = process.communicate()
-        pprint.pprint(report, width=120)
-        print(f"STDOUT: {stdout.decode()}")
-        print(f"STDERR: {stderr.decode()}")
-
-    except Exception as e:
-        import traceback
-
-        traceback.print_exc()
-        import ipdb
-
-        ipdb.set_trace()
-        print(f"Failed to execute command: {str(e)}")
+    pprint.pprint(report, width=120)
```

Comment:
for now ok, but we might not want it to be reported... what would be informative -- point to the files produced (formatted prefix)